### DISCUSSION TITLE
Is it possible to use macros for mongodb collection?

### DISCUSSION THREAD

**Discussion by Wernfried** [syslog-ng/syslog-ng#3875] [Q&A] [answered]
I think about to change my log destination from sql to mongodb. Is it possible to use dynamic collection names, i.e. with macros similar to "file" destination where this is rather common.

Would be like this:

```
destination d_mongodb {
    mongodb(
        uri("mongodb://localhost:27017/syslog")
        collection("msg_${YEAR}-${MONTH}-${DAY}")
        value-pairs(
            scope("selected-macros" "nv-pairs" "sdata")
        )
    );
};

```

Kind Regards
Wernfried

**Comment by szemere** [comment]
Dear @Wernfried ,
 I examined the code. We are currently handling the collection parameter as a string, and passing it to the [mongoc_client_get_collection](http://mongoc.org/libmongoc/1.14.0/mongoc_client_get_collection.html) function, so using macros is not supported currently.

 To apply macros to such a parameter we have to examine the evaluated value per message, and drop the current mongodb **collection**, and get a new one if it has changed. (i.e. **R_SEC** is used.)

 Unfortunately I am not very familiar with mongodb (and it's client library). Maybe you can give me a reference or some advice about the performance impact of collection handling. My first thoughts:
- It might happens that we have to create a new collection for each message. (i.e. hostname macro is used)
- I don't know how many collections can be handled by mongodb. (If someone uses seconds or milliseconds in their collection name, than the number of collections can increase rapidly.)
- How long does it take to allocate and free a collection? It might have a huge performance impact on the mongodb destination.

Thank you for your help!

Br,
Laci

**Comment by Wernfried** [comment]
Hi

I am also not familiar with MongoDB (yet). 

> It might happens that we have to create a new collection for each message. (i.e. hostname macro is used)

This could cause some issues, because you would have to create also indexes and define sharding parameters for load distribution, i.e. adding a collection usually requires some additional administration work.

> I don't know how many collections can be handled by mongodb.

I documentation I found this one:

> The limitation on the number of namespaces is the size of the namespace file divided by 628.
> A 16 megabyte namespace file can support approximately 24,000 namespaces. Each collection and index is a namespace.

So, there is a limit and it would be possible to hit this limit in a real application.

> How long does it take to allocate and free a collection?

For this I did not find anything. Anyway, for me these answers justify that it would be a bad idea to use macros in MongoDB collection name. I will go for a single collection.

Thanks for your reply
Wernfried

**Comment by MrAnno** [comment, marked as answer]
The `collection()` option of the MongoDB destination accepts templates (macros) starting from syslog-ng v3.32.1 (#3621).
